src/search/hnsw_indexer.cc

```diff
@@ -0,0 +1,521 @@
+        current_node.PutMetadata(&current_node_metadata, search_key, storage, batch.Get());
+      }
+
+      entry_points.clear();
+      for (const auto& new_entry_point : nearest_vec_items) {
+        entry_points.push_back(new_entry_point.key);
+      }
+    }
+  } else {
+    auto node = HnswNode(std::string(key), 0);
+    HnswNodeFieldMetadata node_metadata(0, vector);
+    node.PutMetadata(&node_metadata, search_key, storage, batch.Get());
+    metadata->num_levels = 1;
+  }
+
+  while (target_level > metadata->num_levels - 1) {
+    auto node = HnswNode(std::string(key), metadata->num_levels);
+    HnswNodeFieldMetadata node_metadata(0, vector);
+    node.PutMetadata(&node_metadata, search_key, storage, batch.Get());
+    metadata->num_levels++;
+  }
+
+  std::string encoded_index_metadata;
+  metadata->Encode(&encoded_index_metadata);
+  auto index_meta_key = search_key.ConstructFieldMeta();
+  batch->Put(cf_handle, index_meta_key, encoded_index_metadata);
+
+  return Status::OK();
+}
+
+Status HnswIndex::InsertVectorEntry(std::string_view key, const kqir::NumericArray& vector,
+                                    ObserverOrUniquePtr<rocksdb::WriteBatchBase>& batch) {
+  auto target_level = RandomizeLayer();
+  return InsertVectorEntryInternal(key, vector, batch, target_level);
+}
+
+Status HnswIndex::DeleteVectorEntry(std::string_view key, ObserverOrUniquePtr<rocksdb::WriteBatchBase>& batch) const {
+  std::string node_key(key);
+  for (uint16_t level = 0; level < metadata->num_levels; level++) {
+    auto node = HnswNode(node_key, level);
+    auto node_metadata_status = node.DecodeMetadata(search_key, storage);
+    if (!node_metadata_status.IsOK()) {
+      break;
+    }
+
+    auto node_metadata = std::move(node_metadata_status).GetValue();
+    auto node_index_key = search_key.ConstructHnswNode(level, key);
+    auto s = batch->Delete(storage->GetCFHandle(ColumnFamilyID::Search), node_index_key);
+    if (!s.ok()) {
+      return {Status::NotOK, s.ToString()};
+    }
+
+    node.DecodeNeighbours(search_key, storage);
+    for (const auto& neighbour_key : node.neighbours) {
+      GET_OR_RET(RemoveEdge(node_key, neighbour_key, level, batch));
+      auto neighbour_node = HnswNode(neighbour_key, level);
+      HnswNodeFieldMetadata neighbour_node_metadata = GET_OR_RET(neighbour_node.DecodeMetadata(search_key, storage));
+      neighbour_node_metadata.num_neighbours--;
+      neighbour_node.PutMetadata(&neighbour_node_metadata, search_key, storage, batch.Get());
+    }
+  }
+
+  auto has_other_nodes_at_level = [&](uint16_t level, std::string_view skip_key) -> bool {
+    auto prefix = search_key.ConstructHnswLevelNodePrefix(level);
+    util::UniqueIterator it(storage, storage->DefaultScanOptions(), ColumnFamilyID::Search);
+    it->Seek(prefix);
+
+    Slice node_key;
+    Slice node_key_dst;
+    while (it->Valid() && it->key().starts_with(prefix)) {
+      node_key = Slice(it->key().ToString().substr(prefix.size()));
```

Comment:
`it->key().ToString().substr(prefix.size())` is a temporary string object and will be destroyed after this line, and then `node_key` will be dangling.

Reply:
Thank you soooo much!! Really saved me from the persisting problem!! Will make all changes tmr :))